File: iminling-core/src/main/java/com/iminling/core/config/CoreBeanAutoConfiguration.java
package com.iminling.core.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.iminling.common.json.JsonUtil;
import com.iminling.core.ApplicationConstant;
import com.iminling.core.config.argument.DefaultRequestDataReader;
import com.iminling.core.config.argument.GlobalArgumentBeanPostProcessor;
import com.iminling.core.config.argument.GlobalArgumentResolverConfig;
import com.iminling.core.config.exception.GlobalExceptionHandler;
import com.iminling.core.config.mybatis.MybatisConfig;
import com.iminling.core.config.value.GlobalReturnValueHandler;
import com.iminling.core.filter.AuthFilter;
import com.iminling.core.filter.CustomizeGlobalFilter;
import com.iminling.core.filter.Filter;
import com.iminling.core.filter.LoginFilter;
import com.iminling.core.service.ILogService;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;
import org.springframework.core.env.ConfigurableEnvironment;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

import java.util.ArrayList;
import java.util.List;

/**
 * @author yslao@outlook.com
 * @since 2021/2/19
 */
@Import({MybatisConfig.class})
public class CoreBeanAutoConfiguration implements WebMvcConfigurer {

    /*@Bean
    @ConditionalOnMissingBean(name = "defaultAuthFilter")
    public AuthFilter defaultAuthFilter() {
        return new DefaultAuthFilter();
    }

    @Bean
    @ConditionalOnMissingBean(name = "defaultLoginFilter")
    public LoginFilter defaultLoginFilter() {
        return new DefaultLoginFilter();
    }*/

    @Bean
    @ConditionalOnMissingBean(CustomizeGlobalFilter.class)
    public CustomizeGlobalFilter customizeGlobalFilter() {
        return new CustomizeGlobalFilter();
    }

    @Bean
    @ConditionalOnMissingBean(GlobalExceptionHandler.class)
    public GlobalExceptionHandler globalExceptionHandler() {
        return new GlobalExceptionHandler();
    }

    @Bean
    @ConditionalOnMissingBean(GlobalReturnValueHandler.class)
    public GlobalReturnValueHandler globalReturnValueHandler() {
        return new GlobalReturnValueHandler();
    }

    @Bean
    @ConditionalOnMissingBean(DefaultRequestDataReader.class)
    public DefaultRequestDataReader defaultRequestDataReader(ObjectMapper objectMapper) {
        return new DefaultRequestDataReader(objectMapper);
    }

    @Bean
    public ObjectMapper objectMapper() {
        return JsonUtil.getInstant();
    }

    @Bean
    @ConditionalOnMissingBean(GlobalArgumentResolverConfig.class)
    public GlobalArgumentBeanPostProcessor globalArgumentBeanPostProcessor() {
        return new GlobalArgumentBeanPostProcessor();
    }

    /**
     * 生成globalInterceptor
     * @param environment 环境信息
     * @param defaultRequestDataReader 请求数据读取
     * @param loginFilter 登录验证过滤器，有对应bean则size = 0，不会是null
     * @param authFilter 授权验证过滤器，有对应bean则size = 0，不会是null
     * @param logServices 日志记录服务，有对应bean则size = 0，不会是null
     * @return {@link GlobalInterceptor}
     */
    @Bean
    @ConditionalOnMissingBean(GlobalInterceptor.class)
    public GlobalInterceptor globalInterceptor(ConfigurableEnvironment environment,
                                               DefaultRequestDataReader defaultRequestDataReader,
                                               List<LoginFilter> loginFilter, List<AuthFilter> authFilter, List<ILogService> logServices) {
        List<Filter> filters = new ArrayList<>(4);
        String enable = "enable";
        if (environment.getProperty(ApplicationConstant.KEY_FILTER_LOGIN, enable).equals(enable)) {
            filters.addAll(loginFilter);
        }
        if (environment.getProperty(ApplicationConstant.KEY_FILTER_AUTH, enable).equals(enable)) {
            filters.addAll(authFilter);
        }
        if (!environment.getProperty(ApplicationConstant.KEY_FILTER_LOG, enable).equals(enable)) {
            logServices.clear();
        }
        return new GlobalInterceptor(defaultRequestDataReader, filters, logServices, environment);
    }
}

File: iminling-core/src/main/kotlin/com/iminling/core/config/CoreBeanAutoConfiguration.kt
package com.iminling.core.config

import com.fasterxml.jackson.databind.ObjectMapper
import com.iminling.common.http.OkHttpUtils
import com.iminling.common.json.JsonUtil
import com.iminling.core.ApplicationConstant
import com.iminling.core.config.argument.DefaultRequestDataReader
import com.iminling.core.config.argument.GlobalArgumentBeanPostProcessor
import com.iminling.core.config.argument.GlobalArgumentResolverConfig
import com.iminling.core.config.exception.GlobalExceptionHandler
import com.iminling.core.config.filter.AuthFilter
import com.iminling.core.config.filter.Filter
import com.iminling.core.config.filter.LoginFilter
import com.iminling.core.config.jpa.CustomizeJpaConfiguration
import com.iminling.core.config.mybatis.CustomizeMybatisConfig
import com.iminling.core.config.rest.EnhanceRestTemplate
import com.iminling.core.config.rest.RestTemplateErrorHandler
import com.iminling.core.config.rest.TextPlainHttpMessageConverter
import com.iminling.core.config.swagger.SwaggerAutoConfiguration
import com.iminling.core.config.value.GlobalReturnValueHandler
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Import
import org.springframework.core.env.ConfigurableEnvironment
import org.springframework.http.client.OkHttp3ClientHttpRequestFactory
import org.springframework.http.converter.StringHttpMessageConverter
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter
import org.springframework.web.client.RestTemplate
import springfox.documentation.swagger2.annotations.EnableSwagger2WebMvc


/**
 * @author yslao@outlook.com
 * @since 2021/2/19
 */
@EnableSwagger2WebMvc
@Import(CustomizeMybatisConfig::class, CustomizeJpaConfiguration::class, SwaggerAutoConfiguration::class)
class CoreBeanAutoConfiguration {

    /*@Bean
    @ConditionalOnMissingBean(name = "defaultAuthFilter")
    public AuthFilter defaultAuthFilter() {
        return new DefaultAuthFilter();
    }

    @Bean
    @ConditionalOnMissingBean(name = "defaultLoginFilter")
    public LoginFilter defaultLoginFilter() {
        return new DefaultLoginFilter();
    }*/
    @Bean
    @ConditionalOnMissingBean(CustomizeGlobalFilter::class)
    fun customizeGlobalFilter(): CustomizeGlobalFilter {
        return CustomizeGlobalFilter()
    }

    @Bean
    @ConditionalOnMissingBean(GlobalExceptionHandler::class)
    fun globalExceptionHandler(): GlobalExceptionHandler {
        return GlobalExceptionHandler()
    }

    @Bean
    @ConditionalOnMissingBean(GlobalReturnValueHandler::class)
    fun globalReturnValueHandler(): GlobalReturnValueHandler {
        return GlobalReturnValueHandler()
    }

    @Bean
    @ConditionalOnMissingBean(DefaultRequestDataReader::class)
    fun defaultRequestDataReader(objectMapper: ObjectMapper): DefaultRequestDataReader {
        return DefaultRequestDataReader(objectMapper)
    }

    @Bean
    fun objectMapper(): ObjectMapper {
        return JsonUtil.getInstant()
    }

    @Bean
    @ConditionalOnMissingBean(GlobalArgumentResolverConfig::class)
    fun globalArgumentBeanPostProcessor(): GlobalArgumentBeanPostProcessor? {
        return GlobalArgumentBeanPostProcessor()
    }

    /**
     * 生成globalInterceptor
     * @param environment 环境信息
     * @param defaultRequestDataReader 请求数据读取
     * @param loginFilter 登录验证过滤器，有对应bean则size = 0，不会是null
     * @param authFilter 授权验证过滤器，有对应bean则size = 0，不会是null
     * @return [GlobalInterceptor]
     */
    @Bean
    @ConditionalOnMissingBean(GlobalInterceptor::class)
    fun globalInterceptor(
        environment: ConfigurableEnvironment,
        defaultRequestDataReader: DefaultRequestDataReader,
        loginFilter: List<LoginFilter>, authFilter: List<AuthFilter>
    ): GlobalInterceptor {
        val filters: MutableList<Filter> = ArrayList(4)
        val enable = "enable"
        if (environment.getProperty(ApplicationConstant.KEY_FILTER_LOGIN, enable) == enable) {
            filters.addAll(loginFilter)
        }
        if (environment.getProperty(ApplicationConstant.KEY_FILTER_AUTH, enable) == enable) {
            filters.addAll(authFilter)
        }
        return GlobalInterceptor(defaultRequestDataReader, filters, environment)
    }

    @Bean
    @ConditionalOnMissingBean
    fun mappingJackson2HttpMessageConverter(): MappingJackson2HttpMessageConverter {
        return MappingJackson2HttpMessageConverter(JsonUtil.getInstant())
    }

    /**
     * 自定义restTemplate客户端
     * 配置https https://blog.csdn.net/zhousheng193/article/details/84830164
     */
    @Bean
    @ConditionalOnMissingBean(RestTemplate::class)
    fun restTemplate(environment: ConfigurableEnvironment,
                     mappingJackson2HttpMessageConverter: MappingJackson2HttpMessageConverter): RestTemplate {
        val okHttp3ClientHttpRequestFactory = OkHttp3ClientHttpRequestFactory(OkHttpUtils.okHttpClientBuilder().build())
        var enable = environment.getProperty(ApplicationConstant.KEY_REST_TIMEOUT, "false")
        var restTemplate = EnhanceRestTemplate(enable.toBoolean())
        restTemplate.requestFactory = okHttp3ClientHttpRequestFactory
        //val interceptors = restTemplate.interceptors
        //interceptors.add(RestTemplateLoggingInterceptor())
        restTemplate.errorHandler = RestTemplateErrorHandler()
        val messageConverters = restTemplate.messageConverters
        // messageConverters.removeIf { converter: HttpMessageConverter<*>? -> converter is MappingJackson2HttpMessageConverter }
        var listIterator = messageConverters.listIterator()
        while (listIterator.hasNext()) {
            var next = listIterator.next()
            if (next is MappingJackson2HttpMessageConverter) {
                listIterator.set(mappingJackson2HttpMessageConverter)
            }
            if (next is StringHttpMessageConverter) {
                next.defaultCharset = Charsets.UTF_8
            }
        }
        messageConverters.add(TextPlainHttpMessageConverter(JsonUtil.getInstant()))
        return restTemplate
    }

}